src/lib/Sympa/User.pm

```diff
@@ -635,23 +635,45 @@ sub update_global_user {
 
     $who = Sympa::Tools::Text::canonic_email($who);
 
-    ## use hash fingerprint to store password
-    ## hashes that use salts will randomly generate one
-    $values->{'password'} =
-        Sympa::User::password_fingerprint($values->{'password'}, undef)
-        if ($values->{'password'});
-
-    ## Canonicalize lang if possible.
-    $values->{'lang'} = Sympa::Language::canonic_lang($values->{'lang'})
-        || $values->{'lang'}
-        if $values->{'lang'};
 
+    ## Check whether password is already defined.
     my $sdm = Sympa::DatabaseManager->instance;
     unless ($sdm) {
         $log->syslog('err', 'Unavailable database connection');
         return undef;
     }
 
+
+    push @sth_stack, $sth;
+
+    $sth = $sdm->do_query(
+        "SELECT password_user FROM user_table WHERE (email_user=%s)",
+        $sdm->quote($who)
+    );
+    unless (defined $sth) {
+        $log->syslog('err',
+            'Could not check password information for user %s in user_table', $who);
+        $sth = pop @sth_stack;
+        return undef;
+    }
+
+    my $current_password = $sth->fetchrow();
+
+    $sth = pop @sth_stack;
+
+    if ($values->{'password'}) {
+        if($current_password ne $values->{'password'}) {
```

Comment:
This check seems not help for me: `password_user` includes encrypted or hashed value, and it hardly match with new password in clear text form (`$values->{'password'}`).
Consequently, query above to get value of `password_user` field looks not useful.

Reply:
Actually, this test is the whole point of the PR. ;-)
The question is: what should do of a password parameter contained by $values->{'password'}?
1- if it is a new password or a clear text form of an existing password, then we can hash/encrypt it and store in the database,
2- if it is the hashed/encrypted form of an existing password we must not hash/encrypt it again because it would mean replacing the clear text form by the hashed form.
3- IT SHOULD NEVER BE the hashed/encrypted form of a new password, becsause the function assumes that hashing/encrypting must occur here, and not before.

case 1 happens when $current_password ne $values->{'password'}
case 2 happens when $current_password eq $values->{'password'}

Is it clearer this way?

Reply:
* 1 is obviously possible.
  * Is 2 possible? If it is, I think opportunistic solution you say should not be taken. Solution should be that: We correct  the code so that any callers of update_global_user() won't hash/encrypt `$values->{password}`.
  * If 2 is solved, 3 does not matter: Even if user submitted a text looks like hashed/encrypted, it will be treated as legitimate password, and it is encrypted/hashed and stored in table.

----
BTW the way of processing also looks problematic: The way "fetching a row, checking fetched value, then updating the row" is not atomic.

Reply:
Oh yes, 2 is possible: anytime we update an existing user object by passing the sub the content of said user.
That's the reason why the bug was reported after all. That is exactly what happened.
I think that checking data before updating them is not such  a bad idea. And password is the only case in which we have to make such check.

Reply:
I realized that, in the first place, the assumption that password can be updated along with other user attributes is inappropriate.  In fact, there are no cases updating password and other fields simultaneously.  Proper solution may be either:
  - Check content of argument `$value` to construct query updating only available field(s) (cf. `Sympa::List::update_list_member()` and `Sympa::List::update_list_admin()`.
  - Modify `update_global_user()` not to change password field, and add new function to change password, such as `update_user_password()`.